### PR TITLE
Issue 567

### PR DESCRIPTION
**Pull request by yaelal**

### CODE REVIEW

**Comment by GalGend** [comment]
@gruppin @tomers  can you please approve?

**Comment by GalGend** [comment]
@yaelal - can you please next time add a label for the branch name and PR? for example issue_567_hod_hasharon_trees. Thank you so much

**Review by tomers — COMMENTED** [review]
The use of tabs here makes the layout break in some views, like in Github.
I wonder why we even allow tabs in the repo... It's a thing of the past TMO, but that's something religion, so I can understand the may be devs who prefer to have it.

File: server/api/lib/trees/hod_hasharon_tree_permit.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+		}
+		);
+		return treePermits.filter(Boolean); // remove undefined values;
+	}
+	catch (e) {
+		Log.error('error in hod hasharon parse rows:' + e);
+	}
+}
+
+function parseTreesPerPermit(treesInPermitStr, treeAmount) {
```
Comment (tomers) [inline server/api/lib/trees/hod_hasharon_tree_permit.js:102]: Parsing is done on the two parameters side by side. Instead, it should go to helper function

File: server/api/lib/trees/hod_hasharon_tree_permit.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	const linesName = treesInPermitStr.split('\n');
+    const linesAmount = treeAmount.split('\n');
+    var result = {};
+    for (let i = 0; i < linesName.length; ++i) {
+        result[i] = {[linesName[i]]: parseInt(linesAmount[i] || '0')}
+    }
+	return Object.assign({}, ...Object.values(result));
+}
+
+function replaceAll(str, from, to) {
```
Comment (tomers) [inline server/api/lib/trees/hod_hasharon_tree_permit.js:116]: Can we built all these regexp objects as static, not dynamically on function call?

File: server/api/lib/trees/tree_crawler.js [hunk 3]
```diff
@@ -86,14 +87,16 @@ const chooseCrawl = (crawlType) => {
 	const kkl = { 'crawler': crawlTreeExcelByFile, 'permitType': KKLTreePermit };
 	const jer = { 'crawler': crawlTreesHTML , 'permitType': JERTreePermit };
 	const rg  = { 'crawler': crawlRGTreesHTML , 'permitType': RGTreePermit};
+	const hh  = { 'crawler': crawlHHTreesHTML , 'permitType': HHTreePermit};
```
Comment (tomers) [inline server/api/lib/trees/tree_crawler.js:90]: The name `hh` is not obvious. I don't know why we are stinchy when it comes to names. It's hard to read `rg` and say - ah! it is Ramat-Gan. I think we can spare some character for a full name, e.g. `ramat_gan`, `hod_hasharon`.

File: server/api/lib/trees/hod_hasharon_tree_permit.js (hunk 1 above)
Comment (yaelal) [reply server/api/lib/trees/hod_hasharon_tree_permit.js:102]: i dont know what this means

File: server/api/lib/trees/hod_hasharon_tree_permit.js (hunk 2 above)
Comment (yaelal) [reply server/api/lib/trees/hod_hasharon_tree_permit.js:116]: i dont know what this means

File: server/api/lib/trees/tree_crawler.js (hunk 3 above)
Comment (yaelal) [reply server/api/lib/trees/tree_crawler.js]: ok

File: server/api/lib/trees/hod_hasharon_tree_permit.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        const last_date_to_objection = formatDate(raw['ניתן להגיש ערעור עד'], '09:00', 'DD.MM.YY');
+        if (!last_date_to_objection) {
+          Log.error(`No / Bad dates format, ignore this license: hod hasharon, ${raw['שם הרחוב']} , ${raw['ניתן להגיש ערעור עד']}`);
+          return null;
+        }
+        
+        const treesPerPermit = parseTreesPerPermit(raw['סוג העצים'], raw['מספר עצים']);
+        const totalTrees = sum(treesPerPermit);
+        
+        const permitNumber = `meirim-hh-${raw['מספר רישיון']}`;
```
Comment (gruppin) [inline server/api/lib/trees/hod_hasharon_tree_permit.js:69]: קשה לי עם ה hh
אפשר לשנות את זה ל hodash או משהו בסגנון?

Comment (yaelal) [reply server/api/lib/trees/hod_hasharon_tree_permit.js]: changed

File: server/api/lib/trees/hod_hasharon_tree_permit.js (hunk 1 above)
Comment (gruppin) [reply server/api/lib/trees/hod_hasharon_tree_permit.js:102]: he meant that this code snippet code be written as follow (example):
```
function treesStrToArr(str) {
return str.replaceAll('\t', '').replaceAll('\n\n', '\n').split('\n');
}
const linesName = treesStrToArr(treesInPermitStr);
const linesAmount = treesStrToArr(treeAmount);
...
```

File: server/api/lib/trees/hod_hasharon_tree_permit.js (hunk 2 above)
Comment (gruppin) [reply server/api/lib/trees/hod_hasharon_tree_permit.js:116]: no, the from is changing per string. it's a bypass to support replaceAll which exists on Node 16, but not on node 14

File: server/api/lib/trees/hod_hasharon_tree_permit.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        
+        const street = raw['שם הרחוב'];
+        
+        const last_date_to_objection = formatDate(raw['ניתן להגיש ערעור עד'], '09:00', 'DD.MM.YY');
+        if (!last_date_to_objection) {
+          Log.error(`No / Bad dates format, ignore this license: hod hasharon, ${raw['שם הרחוב']} , ${raw['ניתן להגיש ערעור עד']}`);
+          return null;
+        }
+        
+        const treesPerPermit = parseTreesPerPermit(raw['סוג העצים'], raw['מספר עצים']);
```
Comment (tomers) [inline server/api/lib/trees/hod_hasharon_tree_permit.js:66]: Use of Hebrew strings inside code is very problematic. It hurts the maintainability of the code. It hard to edit in some editors.
Can you please maintain a list of constants and use them everywhere you currently have Hebrew string? If I am not mistaken, this is how it is done elsewhere.

File: server/api/lib/trees/hod_hasharon_tree_permit.js (hunk 2 above)
Comment (tomers) [reply server/api/lib/trees/hod_hasharon_tree_permit.js:116]: ack

File: server/api/lib/trees/hod_hasharon_tree_permit.js [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+  catch (e) {
+    Log.error('error in hod hasharon parse rows:' + e);
+  }
+}
+
+function parseTreesPerPermit(treesInPermitStr, treeAmount) {
+    treesInPermitStr = cleanString(treesInPermitStr);
+    treeAmount = cleanString(treeAmount);
+    const linesName = treesInPermitStr.split('\n');
+    const linesAmount = treeAmount.split('\n');
```
Comment (tomers) [inline server/api/lib/trees/hod_hasharon_tree_permit.js:114]: Why not add these lines to `cleanString`?

Comment (yaelal) [reply server/api/lib/trees/hod_hasharon_tree_permit.js:114]: its not cleaning the string
got a new name in mind?

File: server/api/lib/trees/hod_hasharon_tree_permit.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+          [LAST_DATE_TO_OBJECTION]: last_date_to_objection,
+          [PERSON_REQUEST_NAME]: raw['שם בעל הרישיון'],
+          [REASON_SHORT]: raw['סיבת הכריתה'],
+          [TREES_PER_PERMIT]: treesPerPermit,
+          [TOTAL_TREES]: totalTrees,
+          
+          [TREE_PERMIT_URL]: 'https://www.hod-hasharon.muni.il' + raw['url'],
+        };
+        const permit = new TreePermit(attributes);
+        return permit;
```
[the remark this answers is not in the document]
Comment (yaelal) [reply server/api/lib/trees/hod_hasharon_tree_permit.js:95]: that was in the original code, i assumed its done for the const?

File: server/api/lib/trees/hod_hasharon_tree_permit.js (hunk 6 above)
Comment (yaelal) [reply server/api/lib/trees/hod_hasharon_tree_permit.js]: done

File: server/api/lib/trees/hod_hasharon_tree_permit.js (hunk 5 above)
Comment (yaelal) [reply server/api/lib/trees/hod_hasharon_tree_permit.js]: moved

File: server/api/lib/trees/hod_hasharon_tree_permit.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (gruppin) [reply server/api/lib/trees/hod_hasharon_tree_permit.js:95]: it was probably for testing, so i could set the breakpoint and log with conveniece :)

**Pull request merged**